### ISSUE TITLE
Failed to load React due to bundling

### ISSUE THREAD

**Issue by tyh77** [SharePoint/sp-dev-docs#6923]
## Category

- [X] Question
- [ ] Typo
- [ ] Additional article idea

## Question
Was facing React and ReactDOM module loading errors when trying to add React to a non-React SPFx project. I was able to get around the problem by explicitly defining External objects in the webpack configuration in `gulpfile.js` like the following:

```js
build.configureWebpack.mergeConfig({
  additionalConfiguration: (generatedConfiguration) => {
      // remove react, react-dom from Webpack Externals
      generatedConfiguration.externals.splice(generatedConfiguration.externals.indexOf('react'), 1);
      generatedConfiguration.externals.splice(generatedConfiguration.externals.indexOf('react-dom'), 1);

      return generatedConfiguration;
  }
});
```

Without the above configuration, the following error is returned in the local workbench: 
>***Failed to load URL 'https://localhost:4321/node_modules/react/dist/react.js' for resource 'react' in component '802a7ebb-f1c2-4fb7-a5b5-e6e2311b5ab5' (webpart-name). The file was not found in the server. Make sure that you are running 'gulp serve'.

Long story short, this SPFx solution and web part began as a non-React, pure vanilla JS SPFx project. As we're trying to modernize our web part, we're looking to add React into the project and slowly migrate parts of the solution to React.

I compared our solution to a newly created React SPFx web part, and I don't see any externals defined in `config/config.json`, or any additional webpack configurations like the one I'm using in `gulpfile.js`.

**From my understanding, bundling framework is not recommended, but is also not mandatory to ensuring that the web part runs normally. So I'm wondering why my web part is unable to run, at all, if the React and ReactDOM modules are not explicitly excluded from bundling?**

## Environment details (*development & target environment*)
- **Your Developer Environment**: Windows 10
- **Target Environment**: SharePoint Online
- **Framework**: Node.js v10, SPFx v1.10.0
- **Additional details**: react v16.8.5, react-dom v16.8.5

**Comment by andrewconnell** [comment]
React is already excluded from SPFx projects automatically by the SPFx build toolchain... you don't need to exclude it. It's not excluded in the **config.json**, rather it's added to the generated manifest. 

IOW, you don't need to worry about excluding it... SPFx does that automatically because it's already on the SPO pages.

**Comment by tyh77** [comment]
@andrewconnell that's pretty reasonable and it makes complete sense... except in my case - if I didn't explicitly exclude it in the gulpfile, SPFx workbench complains with the error message that I indicated above :(

**Comment by bcameron1231** [comment]
Hi, to confirm, are you trying to add react as an external and not as a linked package in your node_modules?

**Comment by tyh77** [comment]
@bcameron1231 Sorry my mistake, I think I got the concept of the workaround all wrong. From my understanding, those 2 lines in the gulpfile actually remove the react & react-dom from the "externals" array in Webpack - for whatever reason, that is what has worked for me thus far.

For debugging, I removed those gulp configuration commands from my gulpfile. I now get the following error message:

>[SPLoaderError.loadComponentError]:
Failed to load component "802a7ebb-f1c2-4fb7-a5b5-e6e2311b5ab5" (PartnerPortalTestDataGenerationWebPart). Original error: The entry point for component "802a7ebb-f1c2-4fb7-a5b5-e6e2311b5ab5" (PartnerPortalTestDataGenerationWebPart) has a dependency on "react" that is not declared in the manifest.

And adding react to externals has had no effect on eliminating build errors and whatnot.

I also ran `npm ls react` and these are my react dependencies within my SPFx solution:
```
+-- @microsoft/sp-property-pane@1.10.0
| +-- @microsoft/office-ui-fabric-react-bundle@1.10.0
| | `-- react@16.8.5
| `-- react@16.8.5
+-- @microsoft/sp-webpart-base@1.10.0
| `-- @microsoft/sp-loader@1.10.0
|   `-- react@16.8.5
+-- @microsoft/sp-webpart-workbench@1.10.0
| `-- react@16.8.5
+-- react@16.14.0
`-- swagger-ui@3.26.0
  `-- react@15.6.2
```

**Comment by tyh77** [comment]
Adding on to my previous comment, the 2 lines in the gulpfile basically enabled me to load my own version of react, so I'm curious to know why I'm not able to use the existing react from SPFx...

https://github.com/SharePoint/sp-dev-docs/issues/1315#issuecomment-373942024

I'm on React v16.8.5 with SPFx v1.10 - which should be compatible

But I am adding React to a non-React SPFx solution, not sure if that matters

**Comment by andrewconnell** [comment]
@tyh77 said:
> Based on what I currently understand, those 2 lines in the gulpfile actually remove the react & react-dom from the "externals" array in Webpack (please correct me if I'm wrong) - for whatever reason, that is what has worked for me thus far.

it does, but as I said above, it's not needed. That's baked into the SPFx build toolchain.

Those are already added by the toolchain to the webpack config that's dynamically generated when you run **gulp bundle**. You can see this if you dump the webpack config as well as in the generated manifest, as well as if you inspect the bundle from an OOTB React based SPFx project, which doesn't include React.

**Comment by tyh77** [comment]
@andrewconnell gotcha! I looked at the generated manifest and that actually led me straight to the issue with this. I found react listed as a `loaderConfig` object, but the version was for 16.14.0, instead of 16.8.5.

Then I realized that I hadn't been strict in my package.json for react v16.8.5, (had the versioning as `"react": "^16.8.5"`) so NPM naturally found 16.14.0, which is problematic for SPFx v1.10.0.

Making the react & react-dom versions strict allowed the error to go away. Silly error on my end :(
